tuner/libtuner.py

```diff
@@ -0,0 +1,1302 @@
+
+
+def generate_dryrun_dispatch_benchmark_results(
+    compiled_candidates: list[int],
+) -> list[TaskResult]:
+    logging.info("generate_dryrun_dispatch_benchmark_results")
+
+    task_results = [
+        generate_sample_task_result(
+            f"process_time/real_time_mean    {random.uniform(100.0, 500.0):.3g} ms",
+            i,
+            str(0),
+        )
+        for i in compiled_candidates
+    ]
+
+    return task_results
+
+
+def generate_dryrun_model_benchmark_results(
+    model_candidates: list[int],
+) -> tuple[list[TaskResult], list[TaskResult]]:
+    candidate_results = []
+    for i, j in enumerate(model_candidates):
+        stdout = f"process_time/real_time_mean    {random.uniform(100.0, 500.0):.3g} ms"
+        candidate_results.append(generate_sample_task_result(stdout, j, str(i % 3)))
+
+    baseline_results = [
+        generate_sample_task_result(
+            f"process_time/real_time_mean    {random.uniform(100.0, 500.0):.3g} ms",
+            0,
+            str(i),
+        )
+        for i in range(3)
+    ]
+
+    return candidate_results, baseline_results
+
+
+def benchmark_dispatches(
+    args: argparse.Namespace,
+    path_config: PathConfig,
+    compiled_candidates: list[int],
+    candidate_trackers: list[CandidateTracker],
+    tuning_client: TuningClient,
+):
+    logging.info("benchmark_dispatches()")
+
+    if args.dry_run:
+        benchmark_results = generate_dryrun_dispatch_benchmark_results(
+            compiled_candidates
+        )
+    else:
+        # Benchmarking dispatch candidates
+        task_list = [
+            TaskPack(
+                args,
+                candidate_id=i,
+                command=tuning_client.get_dispatch_benchmark_command(
+                    candidate_trackers[i]
+                ),
+                check=False,
+                command_need_device_id=True,
+            )
+            for i in compiled_candidates
+        ]
+        worker_context_queue = create_worker_context_queue(args.devices)
+        benchmark_results = multiprocess_progress_wrapper(
+            num_worker=len(args.devices),
+            task_list=task_list,
+            function=run_command_wrapper,
+            initializer=init_worker_context,
+            initializer_inputs=(worker_context_queue,),
+        )
+
+    (
+        parsed_benchmark_results,
+        dispatch_benchmark_dump_list,
+    ) = parse_dispatch_benchmark_results(
+        path_config, benchmark_results, candidate_trackers
+    )
+    append_to_file(
+        dispatch_benchmark_dump_list,
+        filepath=path_config.output_unilog,
+        title="All Dispatch Benchmark Results",
+    )
+
+    benchmarking_rate = (len(parsed_benchmark_results) / len(benchmark_results)) * 100
+    logging.critical(
+        f"Total: {len(benchmark_results)} | Benchmarked: {len(parsed_benchmark_results)} | Failed: {len(benchmark_results) - len(parsed_benchmark_results)} | Benchmarking Rate: {benchmarking_rate:.1f}%"
+    )
```

Comment:
Are you sure you want `critical` ("A serious error, indicating that the program itself may be unable to continue running.") here? That's a step _above_ error: https://docs.python.org/3/library/logging.html#logging-levels

How about `info`? https://docs.python.org/3/library/logging.html#logging.Logger.info

Reply:
thanks! The original purpose is highlight some important info, and want to log and print those info to the console at the same time. `    # Create a formatter that dynamically adds [levelname] for ERROR and WARNING
    class CustomFormatter(logging.Formatter):
        def format(self, record):
            if record.levelno == logging.INFO:
                return f"{record.message}"
            else:
                return f"[{record.levelname}] {record.message}"`

Now downgrade other `INFO` level to `DEBUG`, and change `CRITICAL` to `INFO`.